Flaky-test: KeySharedSubscriptionTest.testContinueDispatchMessagesWhenMessageDelayed

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Example failures

- [2023-11-04T06:31:52.9828029Z](https://github.com/apache/pulsar/actions/runs/6753154827/job/18359322484#step:9:2318) 


### Exception stacktrace

```
java.lang.AssertionError: expected [80] but found [71]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.failNotEquals(Assert.java:1577)
	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
	at org.testng.Assert.assertEquals(Assert.java:131)
	at org.testng.Assert.assertEquals(Assert.java:1418)
	at org.testng.Assert.assertEquals(Assert.java:1382)
	at org.testng.Assert.assertEquals(Assert.java:1428)
	at org.apache.pulsar.client.api.KeySharedSubscriptionTest.testContinueDispatchMessagesWhenMessageDelayed(KeySharedSubscriptionTest.java:1631)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)

```


### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
@@ -328,17 +328,21 @@ public class PersistentStickyKeyDispatcherMultipleConsumers extends PersistentDi
         acquirePermitsForDeliveredMessages(topic, cursor, totalEntries, totalMessagesSent, totalBytesSent);
 
         // trigger read more messages if necessary
-        if (triggerLookAhead.booleanValue()) {
+        if (triggerLookAhead.booleanValue() && cursor.hasMoreEntries()) {
             // When all messages get filtered and no messages are sent, we should read more entries, "look ahead"
             // so that a possible next batch of messages might contain messages that can be dispatched.
             // This is done only when there's a consumer with available permits, and it's not able to make progress
             // because of blocked hashes. Without this rule we would be looking ahead in the stream while the
             // new consumers are not ready to accept the new messages,
             // therefore would be most likely only increase the distance between read-position and mark-delete position.
+            // Look-ahead is engaged only when the cursor has more entries. Otherwise the next readMoreEntries call
+            // would skip replaying the replay queue and pulling due messages from the delayed delivery tracker, and
+            // instead issue a normal read that waits at the end of the topic for new entries. That would leave
+            // deliverable messages stuck in the replay queue or the delayed delivery tracker until an unrelated
+            // event (such as a consumer flow request) triggers another read, stalling dispatch (issue #21554).
             skipNextReplayToTriggerLookAhead = true;
             // skip backoff delay before reading ahead in the "look ahead" mode to prevent any additional latency
-            // only skip the delay if there are more entries to read
-            skipNextBackoff = cursor.hasMoreEntries();
+            skipNextBackoff = true;
             return true;
         }
 

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
@@ -328,21 +328,28 @@ public class PersistentStickyKeyDispatcherMultipleConsumers extends PersistentDi
         acquirePermitsForDeliveredMessages(topic, cursor, totalEntries, totalMessagesSent, totalBytesSent);
 
         // trigger read more messages if necessary
-        if (triggerLookAhead.booleanValue() && cursor.hasMoreEntries()) {
+        if (triggerLookAhead.booleanValue() && (allowOutOfOrderDelivery || cursor.hasMoreEntries())) {
             // When all messages get filtered and no messages are sent, we should read more entries, "look ahead"
             // so that a possible next batch of messages might contain messages that can be dispatched.
             // This is done only when there's a consumer with available permits, and it's not able to make progress
             // because of blocked hashes. Without this rule we would be looking ahead in the stream while the
             // new consumers are not ready to accept the new messages,
             // therefore would be most likely only increase the distance between read-position and mark-delete position.
-            // Look-ahead is engaged only when the cursor has more entries. Otherwise the next readMoreEntries call
-            // would skip replaying the replay queue and pulling due messages from the delayed delivery tracker, and
-            // instead issue a normal read that waits at the end of the topic for new entries. That would leave
-            // deliverable messages stuck in the replay queue or the delayed delivery tracker until an unrelated
-            // event (such as a consumer flow request) triggers another read, stalling dispatch (issue #21554).
+            // When ordered delivery is required, look-ahead is engaged only when the cursor has more entries.
+            // Otherwise the next readMoreEntries call would skip replaying the replay queue and pulling due messages
+            // from the delayed delivery tracker, and instead issue a normal read that waits at the end of the topic
+            // for new entries. That would leave deliverable messages stuck in the replay queue or the delayed
+            // delivery tracker until an unrelated event (such as a consumer flow request) triggers another read,
+            // stalling dispatch (issue #21554).
+            // When out-of-order delivery is allowed, look-ahead is engaged unconditionally, as before. In that mode
+            // the replay queue doesn't track sticky key hashes, so the replay position filter cannot exclude
+            // messages for consumers without available permits, and each replay would re-read and discard the same
+            // undispatchable messages. Ending the cycle with a look-ahead attempt (that waits at the end of the
+            // topic when there is nothing to read) prevents such repeated read-and-discard loops.
             skipNextReplayToTriggerLookAhead = true;
             // skip backoff delay before reading ahead in the "look ahead" mode to prevent any additional latency
-            skipNextBackoff = true;
+            // only skip the delay if there are more entries to read
+            skipNextBackoff = cursor.hasMoreEntries();
             return true;
         }
 

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumersTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumersTest.java
@@ -479,6 +479,105 @@ public class PersistentStickyKeyDispatcherMultipleConsumersTest {
         allEntries.forEach(Entry::release);
     }
 
+    /**
+     * Reproduces the dispatch stall behind the flaky
+     * KeySharedSubscriptionTest.testContinueDispatchMessagesWhenMessageDelayed (issue #21554).
+     *
+     * When a replay read gets fully discarded (for example because the target consumer ran out of permits while
+     * the read was in flight) and the cursor has no more entries, engaging the "look ahead" mode made the follow-up
+     * read skip the replay queue and issue a normal read that waits at the end of the topic for new entries. Messages
+     * in the replay queue for consumers with available permits were then stuck until an unrelated event, such as a
+     * consumer flow request, triggered another read.
+     */
+    @Test(timeOut = 30000)
+    public void testLookAheadNotEngagedWhenCursorHasNoMoreEntries() throws Exception {
+        persistentDispatcher.close();
+
+        // the mocked executor doesn't support schedule(), so run the rescheduled read directly
+        persistentDispatcher = new PersistentStickyKeyDispatcherMultipleConsumers(
+                topicMock, cursorMock, subscriptionMock, configMock,
+                new KeySharedMeta().setKeySharedMode(KeySharedMode.AUTO_SPLIT)) {
+            @Override
+            protected void reScheduleReadInMs(long readAfterMs) {
+                orderedExecutor.execute(this::readMoreEntries);
+            }
+        };
+
+        // consumer1 has available permits at all times
+        persistentDispatcher.addConsumer(consumerMock).join();
+
+        // the slow consumer initially has 2 permits
+        final Consumer slowConsumerMock = createMockConsumer();
+        doReturn("consumer2").when(slowConsumerMock).consumerName();
+        doReturn(true).when(slowConsumerMock).isWritable();
+        AtomicInteger slowConsumerAvailablePermits = new AtomicInteger(2);
+        doAnswer(invocation -> slowConsumerAvailablePermits.get()).when(slowConsumerMock).getAvailablePermits();
+        persistentDispatcher.addConsumer(slowConsumerMock).join();
+
+        StickyKeyConsumerSelector selector = persistentDispatcher.getSelector();
+        String keyForConsumer1 = generateKeyForConsumer(selector, consumerMock);
+        String keyForSlowConsumer = generateKeyForConsumer(selector, slowConsumerMock);
+
+        final Entry entry1 = createEntry(1, 1, "message1", 1, keyForSlowConsumer);
+        final Entry entry2 = createEntry(1, 2, "message2", 2, keyForSlowConsumer);
+        final Entry entry3 = createEntry(1, 3, "message3", 3, keyForConsumer1);
+        final List<Entry> allEntries = List.of(entry1, entry2, entry3);
+
+        // the cursor has no more entries; a normal read would wait for new entries without completing
+        doReturn(false).when(cursorMock).hasMoreEntries();
+        doAnswer(invocationOnMock -> null)
+                .when(cursorMock).asyncReadEntriesWithSkipOrWait(anyInt(), anyLong(), any(), any(), any(), any());
+
+        // Mock Cursor#asyncReplayEntries. While the first replay read is in flight, the slow consumer runs out of
+        // permits and a message for consumer1 becomes replayable. This mirrors the delayed delivery tracker feeding
+        // messages to the replay queue while dispatching is in progress.
+        doAnswer(invocationOnMock -> {
+            Set<Position> positionsArg = invocationOnMock.getArgument(0);
+            Set<Position> positions = new TreeSet<>(positionsArg);
+            if (!positions.contains(entry3.getPosition())) {
+                slowConsumerAvailablePermits.set(0);
+                // add extra retain since addEntryToReplay will release it
+                ((EntryImpl) entry3).retain();
+                persistentDispatcher.addEntryToReplay(entry3);
+            }
+            List<Entry> entries = allEntries.stream()
+                    .filter(entry -> positions.contains(entry.getPosition()))
+                    .toList();
+            AsyncCallbacks.ReadEntriesCallback callback = invocationOnMock.getArgument(1);
+            Object ctx = invocationOnMock.getArgument(2);
+            callback.readEntriesComplete(copyEntries(entries), ctx);
+            return Collections.emptySet();
+        }).when(cursorMock).asyncReplayEntries(anySet(), any(), any(), anyBoolean());
+
+        CountDownLatch consumer1ReceivedMessage3 = new CountDownLatch(1);
+        mockSendMessages(consumerMock, entries -> {
+            boolean message3Found = entries.stream()
+                    .anyMatch(entry -> entry.getPosition().equals(entry3.getPosition()));
+            if (message3Found) {
+                consumer1ReceivedMessage3.countDown();
+            }
+        });
+        mockSendMessages(slowConsumerMock, entries -> { });
+
+        // seed the replay queue with the slow consumer's entries
+        for (Entry entry : List.of(entry1, entry2)) {
+            // add extra retain since addEntryToReplay will release it
+            ((EntryImpl) entry).retain();
+            persistentDispatcher.addEntryToReplay(entry);
+        }
+
+        // trigger the replay read. The batch gets fully discarded since the slow consumer runs out of permits while
+        // the read is in flight. The message for consumer1 that was added to the replay queue in the meantime must
+        // get dispatched by the follow-up read instead of the dispatcher parking a normal read at the end of the
+        // topic.
+        persistentDispatcher.readMoreEntries();
+
+        assertTrue(consumer1ReceivedMessage3.await(5, TimeUnit.SECONDS),
+                "The replayed message for consumer1 with available permits should have been dispatched");
+
+        allEntries.forEach(Entry::release);
+    }
+
     @Test(timeOut = 30000)
     public void testMessageRedelivery() throws Exception {
         final List<Position> actualEntriesToConsumer1 = new CopyOnWriteArrayList<>();

